subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java

```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
+        }
+        Set<String> consumedLocations = new LinkedHashSet<>();
+        node.getTaskProperties().getInputFileProperties()
+            .forEach(spec -> spec.getPropertyFiles().visitStructure(new FileCollectionStructureVisitor() {
+                @Override
+                public void visitCollection(FileCollectionInternal.Source source, Iterable<File> contents) {
+                    contents.forEach(location -> consumedLocations.add(location.getAbsolutePath()));
+                }
+
+                @Override
+                public void visitGenericFileTree(FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                    fileTree.forEach(location -> consumedLocations.add(location.getAbsolutePath()));
+                }
+
+                @Override
+                public void visitFileTree(File root, PatternSet patterns, FileTreeInternal fileTree) {
+                    consumedLocations.add(root.getAbsolutePath());
+                }
+
+                @Override
+                public void visitFileTreeBackedByFile(File file, FileTreeInternal fileTree, FileSystemMirroringFileTree sourceTree) {
+                    consumedLocations.add(file.getAbsolutePath());
+                }
+            }));
+        consumedDirectories.recordRelatedToNode(node, consumedLocations);
+        for (String consumedLocation : consumedLocations) {
+            consumedAndProducedLocations.getProducedDirectories().getNodesRelatedTo(consumedLocation).stream()
+                .filter(producerNode -> missesDependency(producerNode, node))
+                .forEach(producerWithoutDependency -> emitMissingDependencyDeprecationWarning(producerWithoutDependency, node));
+        }
+    }
+
+    private static boolean missesDependency(Node producer, Node consumer) {
+        if (consumer == producer) {
+            return false;
+        }
+        if (consumer.getDependencySuccessors().contains(producer)) {
+            return false;
+        }
+        for (Node dependency : consumer.getAllSuccessors()) {
+            if (dependency == producer || dependency.getDependencySuccessors().contains(producer)) {
+                return false;
+            }
+        }
+        // Do a deep search
+        ArrayDeque<Node> queue = new ArrayDeque<>();
+        consumer.getAllSuccessors().forEach(queue::add);
+        while (!queue.isEmpty()) {
+            Node dependency = queue.removeFirst();
+            if (dependency == producer) {
+                return false;
+            }
+            dependency.getAllSuccessors().forEach(queue::add);
+        }
+        return true;
+    }
+
+    private void emitMissingDependencyDeprecationWarning(Node producer, Node consumer) {
+        DeprecationLogger.deprecateBehaviour(String.format("%s consumes the output of %s, but does not declare a dependency.", consumer, producer)).willBeRemovedInGradle7().undocumented().nagUser();
```

Comment:
I think these calls are a lot more readable if you put each `.xxx()` call on a new line.

Reply:
The deprecation logger call is gone now.